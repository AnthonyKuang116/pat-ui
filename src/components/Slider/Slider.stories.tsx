import React from 'react';
import {action} from '@storybook/addon-actions';
import Slider from './Slider';

export default {
    title: 'Slider',
    component: Slider,
};

export const DefaultSlider = () => (
    <Slider onChange={action("Thumb slider was moved")}/>
)

export const DiffSizeSlider = () => (
    <div>
        <Slider
            sliderSize='sm'
            onChange={action("Small slider was moved")}
        />
        <Slider
            sliderSize='md'
            onChange={action("Medium slider was moved")}
        />
        <Slider
            sliderSize='lg'
            onChange={action("Large slider was moved")}
        />
    </div>
)

export const DiffSizeThumb = () => (
    <div>
        <Slider
            thumbSize='sm'
            onChange={action("Small thumb slider was moved")}
        />
        <Slider
            thumbSize='md'
            onChange={action("Medium thumb slider was moved")}
        />
        <Slider
            thumbSize='lg'
            onChange={action("Large thumb slider was moved")}
        />
    </div>
)

export const DiffOrientation = () => (
    <div style={{height: "7rem", display: "flex", alignItems: "center"}}>
        <Slider
            sliderOrientation='horizontal'
            onChange={action("Horiztonal slider was moved")}
        />
        <Slider
            sliderOrientation='vertical'
            onChange={action("Vertical slider was moved")}
        />
    </div>
)

export const DiffTypeSlider = () => (
    <div>
        <Slider
            sliderTheme='primary'
            onChange={action("Primary slider was moved")}
        />
        <Slider 
            sliderTheme='secondary'
            onChange={action("Secondary slider was moved")}
        />
        <Slider 
            sliderTheme='success'
            onChange={action("Success slider was moved")}
        />
        <Slider 
            sliderTheme='info'
            onChange={action("Info slider was moved")}
        />
        <Slider 
            sliderTheme='warning'
            onChange={action("Warning slider was moved")}
        />
        <Slider 
            sliderTheme='danger'
            onChange={action("Danger slider was moved")}
        />
        <Slider 
            sliderTheme='light'
            onChange={action("Light slider was moved")}
        />
        <Slider 
            sliderTheme='dark'
            onChange={action("Dark slider was moved")}
        />
    </div>
)

export const DiffTypeThumb = () => (
    <div>
        <Slider
            thumbTheme='primary'
            onChange={action("Primary thumb was moved")}
        />
        <Slider 
            thumbTheme='secondary'
            onChange={action("Secondary thumb was moved")}
        />
        <Slider 
            thumbTheme='success'
            onChange={action("Success thumb was moved")}
        />
        <Slider 
            thumbTheme='info'
            onChange={action("Info thumb was moved")}
        />
        <Slider 
            thumbTheme='warning'
            onChange={action("Warning thumb was moved")}
        />
        <Slider 
            thumbTheme='danger'
            onChange={action("Danger thumb was moved")}
        />
        <Slider 
            thumbTheme='light'
            onChange={action("Light thumb was moved")}
        />
        <Slider 
            thumbTheme='dark'
            onChange={action("Dark thumb was moved")}
        />
    </div>
)

