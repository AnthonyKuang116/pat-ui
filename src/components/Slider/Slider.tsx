import React, { FC, useState } from 'react';
import { classNames } from '../../utils/classNames';

export type SliderSize = 'lg' | 'md' | 'sm';
export type ThumbSize = 'lg' | 'md' | 'sm';
export type SliderOrientation = 'horizontal' | 'vertical';
export type SliderTheme =
  | 'primary'
  | 'secondary'
  | 'success'
  | 'info'
  | 'warning'
  | 'danger'
  | 'light'
  | 'dark';
export type ThumbTheme =
  | 'primary'
  | 'secondary'
  | 'success'
  | 'info'
  | 'warning'
  | 'danger'
  | 'light'
  | 'dark';

export interface ISliderProps {
  /**set slider size */
  sliderSize?: SliderSize;
  /**set thumb size */
  thumbSize?: ThumbSize;
  /**set customized slider */
  className?: string;
  /**set slider orientation */
  sliderOrientation?: SliderOrientation;
  /**set slider type */
  sliderTheme?: SliderTheme;
  /**set thumb type */
  thumbTheme?: ThumbTheme;
  /**set min slider range */
  min?: number;
  /**set max slider range */
  max?: number;
  /**set inital starting value of thumb on slider */
  startvalue?: number;
  /**set action onChange */
  onChange?: Function;
}

export type PatSliderProps = ISliderProps;

export const Slider: FC<PatSliderProps> = (props) => {
  const {min, max, className, sliderSize, thumbSize, sliderTheme, thumbTheme, sliderOrientation, onChange} = props
  const [value, setValue] = useState(props.startvalue);
  
  let styleClasses = classNames('slider', {
    [`slider-${sliderSize}`]: !!sliderSize,
    [`thumb-${thumbSize}`]: !!thumbSize,
    [`thumb-${thumbTheme}`]: true,
    [`slider-${sliderTheme}`]: true,
    [`slider-${sliderOrientation}`]: true
  });
  if (className) {
    styleClasses += ' ' + className;
  }

  let slider = (
    <input
      data-testid='slider-element'
      type="range"
      min={min}
      max={max}
      className={styleClasses}
      value={value}
      onChange={(event:React.ChangeEvent<HTMLInputElement>) => {
        setValue(+event.target.value);
        if (onChange) {
          onChange(value);
        }
      }}
    ></input>
  );

  return slider;
};

Slider.defaultProps = {
  sliderTheme: 'secondary',
  thumbTheme: 'success',
  sliderOrientation: 'horizontal',
  sliderSize: 'md',
  thumbSize: 'md',
  startvalue: 0,
  min: 0,
  max: 100,
};

export default Slider;
