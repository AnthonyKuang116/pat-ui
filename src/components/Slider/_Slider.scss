.slider {
  -webkit-appearance: none;
  outline: none;
  opacity: 0.7;
  -webkit-transition: 0.2s;
  transition: opacity 0.2s;
  border-radius: 15px;

  &::-moz-range-thumb {
    border-radius: 15px;
    cursor: pointer;
  }
  &::-webkit-slider-thumb {
    -webkit-appearance: none;
    border-radius: 15px;
    cursor: pointer;
  }
}

//thumb colors
.thumb-primary {
  &::-moz-range-thumb{
    @include thumb-style($primary);
  }
  &::-webkit-slider-thumb {
    @include thumb-style($primary);
  }
}
.thumb-secondary {
  &::-moz-range-thumb{
    @include thumb-style($secondary);
  }
  &::-webkit-slider-thumb {
    @include thumb-style($secondary);
  }
}
.thumb-success {
  &::-moz-range-thumb{
    @include thumb-style($success);
  }
  &::-webkit-slider-thumb {
    @include thumb-style($success);
  }
}
.thumb-info {
  &::-moz-range-thumb{
    @include thumb-style($info);
  }
  &::-webkit-slider-thumb {
    @include thumb-style($info);
  }
}
.thumb-warning {
  &::-moz-range-thumb{
    @include thumb-style($warning);
  }
  &::-webkit-slider-thumb {
    @include thumb-style($warning);
  }
}
.thumb-danger{
  &::-moz-range-thumb{
    @include thumb-style($danger);
  }
  &::-webkit-slider-thumb {
    @include thumb-style($danger);
  }
}
.thumb-light {
  &::-moz-range-thumb{
    @include thumb-style($light);
  }
  &::-webkit-slider-thumb {
    @include thumb-style($light);
  }
}
.thumb-dark {
  &::-moz-range-thumb{
    @include thumb-style($dark);
  }
  &::-webkit-slider-thumb {
    @include thumb-style($dark);
  }
}

//slider sizes
.slider-sm {
  @include slider-size($slider-height-sm);
}
.slider-md {
  @include slider-size($slider-height-md);
}
.slider-lg {
  @include slider-size($slider-height-lg);
}

//thumb sizes
.thumb-sm{
  &::-moz-range-thumb{
    @include thumb-size($thumb-width-sm, $thumb-height-sm);
  }
  &::-webkit-slider-thumb {
    @include thumb-size($thumb-width-sm, $thumb-height-sm);
  }
}
.thumb-md{
  &::-moz-range-thumb{
    @include thumb-size($thumb-width-md, $thumb-height-md);
  }
  &::-webkit-slider-thumb {
    @include thumb-size($thumb-width-md, $thumb-height-md);
  }
}
.thumb-lg{
  &::-moz-range-thumb{
    @include thumb-size($thumb-width-lg, $thumb-height-lg);
  }
  &::-webkit-slider-thumb {
    @include thumb-size($thumb-width-lg, $thumb-height-lg);
  }
}

//slider colors
.slider-primary {
  @include slider-style($primary, #19dfa4);
}
.slider-secondary {
  @include slider-style($secondary, #9b9fa4);
}
.slider-success {
  @include slider-style($success, #63f21d);
}
.slider-info {
  @include slider-style($info, #16d7f5);
}
.slider-warning {
  @include slider-style($warning, #fff08f)
}
.slider-danger{
  @include slider-style($danger, #fc0a21)
}
.slider-light{
  @include slider-style($light, #b9b9b9)
}
.slider-dark{
  @include slider-style($dark, #65686b)
}
