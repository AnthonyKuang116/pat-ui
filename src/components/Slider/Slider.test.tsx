import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Slider, { PatSliderProps } from './Slider';

describe('Slider', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<Slider />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render default slider', () => {
    const sliderProps: PatSliderProps = {
      sliderTheme: 'secondary',
      thumbTheme: 'success',
      sliderOrientation: 'horizontal',
      sliderSize: 'md',
      thumbSize: 'md',
      startvalue: 0,
      min: 0,
      max: 0,
      onChange: jest.fn(),
    };
    const wrapper = render(<Slider {...sliderProps}></Slider>);
    const element = wrapper.getByTestId('slider-element') as HTMLElement;

    expect(element).toBeInTheDocument();
    expect(element).toHaveClass('slider');
    expect(sliderProps.onChange).toHaveBeenCalledTimes(0);
    fireEvent.change(element, { target: { value: 1 } });
    expect(sliderProps.onChange).toHaveBeenCalledTimes(1);
    expect(element.tagName).toBe('INPUT');
    expect(element).toHaveClass(
      'slider slider-secondary thumb-success slider-horizontal slider-md thumb-md'
    );
  });

  it('should render the correct classNames', () => {
    const sliderProps: PatSliderProps = {
      sliderSize: 'sm',
      sliderTheme: 'primary',
      className: 'test',
    };
    const wrapper = render(<Slider {...sliderProps}></Slider>);
    const element = wrapper.getByTestId('slider-element') as HTMLElement;
    expect(element).toBeInTheDocument();
    expect(element).toHaveClass('slider test');
    expect(element.tagName).toBe('INPUT');
  });

  it('should render a slider with the correct props', () => {
    const sliderProps: PatSliderProps = {
      sliderSize: 'sm',
      thumbSize: 'lg',
      className: 'test',
      sliderOrientation: 'vertical',
      sliderTheme: 'info',
      thumbTheme: 'primary',
    };
    const wrapper = render(<Slider {...sliderProps} />);
    const element = wrapper.getByTestId('slider-element') as HTMLElement;

    expect(element).toBeInTheDocument();
    expect(element.tagName).toBe('INPUT');
    expect(element).toHaveClass(
      'slider slider-sm thumb-lg test slider-vertical slider-info thumb-primary'
    );
  });

  it('should render a slider with 2 inputs from user', () => {
    const sliderProps: PatSliderProps = {
      onChange: jest.fn(),
    };
    const wrapper = render(<Slider {...sliderProps}></Slider>);
    const element = wrapper.getByTestId('slider-element') as HTMLElement;

    expect(element).toBeInTheDocument();
    expect(element.tagName).toBe('INPUT');
    expect(sliderProps.onChange).toHaveBeenCalledTimes(0);
    fireEvent.change(element, { target: { value: 1 } });
    expect(sliderProps.onChange).toHaveBeenCalledTimes(1);
    fireEvent.change(element, { target: { value: 2 } });
    expect(sliderProps.onChange).toHaveBeenCalledTimes(2);
  });
});
