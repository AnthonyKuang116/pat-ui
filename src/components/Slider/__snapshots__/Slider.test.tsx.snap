// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider should match snapshot 1`] = `
<DocumentFragment>
  <input
    class="slider slider-md thumb-md thumb-success slider-secondary slider-horizontal"
    data-testid="slider-element"
    max="100"
    min="0"
    type="range"
    value="0"
  />
</DocumentFragment>
`;
